File: packages/gbnf-experiment/data/20260909T130706Z_31ac2789/ported_implementation/src/grammar_graph/graph_pointer.ts
import { printGraphPointer } from './print.ts';
import {
  isGraphPointerRuleChar,
  isGraphPointerRuleCharExclude,
  isGraphPointerRuleEnd,
  isGraphPointerRuleRef,
  isRuleEnd,
} from './type_guards.ts';

import type {
  PrintOpts,
  ResolvedGraphPointer,
  UnresolvedRule,
} from './grammar_graph_types.ts';
import type { GraphNode } from './graph_node.ts';

export type GraphPointerKey = string;

export class GraphPointer<T extends UnresolvedRule = UnresolvedRule> {
  node: GraphNode<T>;
  parent: GraphPointer | undefined;
  id: string;

  private __valid__: boolean | undefined = undefined;

  constructor(node: GraphNode<T>, parent?: GraphPointer) {
    if (node === undefined || node === null) {
      throw new Error('Node is undefined');
    }
    this.node = node;
    this.parent = parent ?? undefined;
    this.id = parent ? `${parent.id}-${node.id}` : node.id;
  }

  get rule(): T {
    return this.node.rule;
  }

  get valid(): boolean | undefined {
    return this.__valid__;
  }

  set valid(valid: boolean | undefined) {
    this.__valid__ = valid;
  }

  print(opts: PrintOpts): string {
    return printGraphPointer(this as GraphPointer)(opts);
  }

  /**
   * Resolve the graph pointer, yielding pointers that point at a concrete
   * (non reference) rule.
   */
  *resolve(resolved = false): Generator<ResolvedGraphPointer> {
    const pointer = this as GraphPointer;
    if (isGraphPointerRuleRef(pointer)) {
      if (resolved) {
        if (!pointer.node.next) {
          throw new Error(`No next node: ${pointer.node.id}`);
        }
        yield* new GraphPointer(pointer.node.next, pointer.parent).resolve();
      } else {
        for (const node of pointer.rule.nodes) {
          yield* new GraphPointer(node, pointer).resolve();
        }
      }
    } else if (isGraphPointerRuleEnd(pointer)) {
      if (!pointer.parent) {
        yield pointer as ResolvedGraphPointer;
      } else {
        yield* pointer.parent.resolve(true);
      }
    } else if (isGraphPointerRuleChar(pointer) || isGraphPointerRuleCharExclude(pointer)) {
      yield pointer as ResolvedGraphPointer;
    } else {
      throw new Error(`Unknown rule: ${pointer.node.rule}`);
    }
  }

  /**
   * Fetch the next resolved graph pointers.
   */
  *fetchNext(): Generator<ResolvedGraphPointer> {
    // If this pointer is invalid, then we don't return any new pointers
    if (!this.valid) {
      return;
    }

    // If this pointer is an end node, we return the parent's next node.
    // If no parent exists, we return nothing, since it's the end of the line.
    if (isRuleEnd(this.node.rule)) {
      if (this.parent) {
        yield* this.parent.fetchNext();
      }
    } else {
      if (!this.node.next) {
        throw new Error(`No next node: ${this.node.id}`);
      }
      const pointer = new GraphPointer(this.node.next, this.parent);
      yield* pointer.resolve();
    }
  }

  fetch_next(): Generator<ResolvedGraphPointer> {
    return this.fetchNext();
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T130706Z_31ac2789/20260910T073404Z_10c90cd6/ported_implementation/gbnf/grammar_graph/graph_pointer.py
from typing import Any, Dict, Generator, Optional

from .grammar_graph_types import UnresolvedRule
from .graph_node import GraphNode
from .print import print_graph_pointer
from .type_guards import (
    is_graph_pointer_rule_char,
    is_graph_pointer_rule_char_exclude,
    is_graph_pointer_rule_end,
    is_graph_pointer_rule_ref,
    is_rule_end,
)

GraphPointerKey = str


class GraphPointer:
    def __init__(self, node: GraphNode, parent: Optional["GraphPointer"] = None):
        if node is None:
            raise ValueError("Node is undefined")
        self.node = node
        self.parent = parent
        self.id = f"{parent.id}-{node.id}" if parent else node.id
        self.__valid__: Optional[bool] = None

    @property
    def rule(self) -> UnresolvedRule:
        return self.node.rule

    @property
    def valid(self) -> Optional[bool]:
        return self.__valid__

    @valid.setter
    def valid(self, valid: Optional[bool]) -> None:
        self.__valid__ = valid

    def print(self, opts: Dict[str, Any]) -> str:
        return print_graph_pointer(self)(opts)

    def resolve(self, resolved: bool = False) -> Generator["GraphPointer", None, None]:
        """Resolve the graph pointer, yielding pointers that point at a
        concrete (non reference) rule.
        """
        pointer = self
        if is_graph_pointer_rule_ref(pointer):
            if resolved:
                if not pointer.node.next:
                    raise ValueError(f"No next node: {pointer.node.id}")
                yield from GraphPointer(pointer.node.next, pointer.parent).resolve()
            else:
                for node in pointer.rule.nodes:
                    yield from GraphPointer(node, pointer).resolve()
        elif is_graph_pointer_rule_end(pointer):
            if not pointer.parent:
                yield pointer
            else:
                yield from pointer.parent.resolve(True)
        elif is_graph_pointer_rule_char(pointer) or is_graph_pointer_rule_char_exclude(
            pointer
        ):
            yield pointer
        else:
            raise ValueError(f"Unknown rule: {pointer.node.rule}")

    def fetch_next(self) -> Generator["GraphPointer", None, None]:
        """Fetch the next resolved graph pointers."""
        # If this pointer is invalid, then we don't return any new pointers
        if not self.valid:
            return

        # If this pointer is an end node, we return the parent's next node.
        # If no parent exists, we return nothing, since it's the end of the line.
        if is_rule_end(self.node.rule):
            if self.parent:
                yield from self.parent.fetch_next()
        else:
            if not self.node.next:
                raise ValueError(f"No next node: {self.node.id}")
            pointer = GraphPointer(self.node.next, self.parent)
            yield from pointer.resolve()

    fetchNext = fetch_next

    def __repr__(self) -> str:
        return f"GraphPointer({self.id})"
